File: Tests/Conformance/AudioDeviceFacadePrototype05827.zig
const std = @import("std");
const r4os = @import("r4os");

var sys_table: r4os.abi.R4XStartR4Sys = .{};
var audio_table: r4os.abi.R4XStartR4Audio = .{};
var dev_table: r4os.abi.R4XStartR4Dev = .{};
var context: r4os.abi.R4XStartContext = .{};
var timeout_next = false;
var stream_open = false;
var open_connections: u32 = 0;

fn fakeServiceOpen(_: [*:0]const u8, out: *r4os.abi.ServiceInfo) callconv(.c) i32 {
    out.* = .{ .handle = 77 };
    open_connections += 1;
    return r4os.abi.service_api_result_ok;
}

fn fakeServiceClose(handle: u32) callconv(.c) i32 {
    if (handle != 77 or open_connections == 0) return r4os.abi.service_api_result_bad_handle;
    open_connections -= 1;
    return r4os.abi.service_api_result_ok;
}

fn copyResponse(response: [*]u8, capacity: u32, value: *const r4os.abi.AudioServiceStreamResult) i32 {
    if (capacity < @sizeOf(r4os.abi.AudioServiceStreamResult)) return r4os.abi.service_api_result_buffer_too_small;
    @memcpy(response[0..@sizeOf(r4os.abi.AudioServiceStreamResult)], std.mem.asBytes(value));
    return @sizeOf(r4os.abi.AudioServiceStreamResult);
}

fn fakeServiceCall(_: u32, op: u16, request: [*]const u8, request_len: u32, header: *r4os.abi.ServiceMessageHeader, response: [*]u8, capacity: u32, _: u64) callconv(.c) i32 {
    header.* = .{ .op = op, .status = r4os.abi.service_api_result_ok };
    if (timeout_next) {
        timeout_next = false;
        return r4os.abi.service_api_result_timeout;
    }
    var result = r4os.abi.AudioServiceStreamResult{ .action = op, .result = 0, .stream_id = 91 };
    switch (op) {
        r4os.abi.audio_service_op_open_stream => stream_open = true,
        r4os.abi.audio_service_op_write_stream => {
            if (!stream_open or request_len < @sizeOf(r4os.abi.AudioServiceStreamWriteRequest)) result.result = -1 else {
                var wire: r4os.abi.AudioServiceStreamWriteRequest = .{};
                @memcpy(std.mem.asBytes(&wire), request[0..@sizeOf(r4os.abi.AudioServiceStreamWriteRequest)]);
                result.bytes = wire.byte_count;
                result.result = @intCast(wire.byte_count);
            }
        },
        r4os.abi.audio_service_op_set_stream_volume => if (!stream_open) {
            result.result = -1;
        },
        r4os.abi.audio_service_op_close_stream => if (stream_open) {
            stream_open = false;
        } else {
            result.result = -1;
        },
        else => result.result = r4os.abi.service_api_result_bad_op,
    }
    return copyResponse(response, capacity, &result);
}

fn fakeAudioOpen(_: u32, _: u16, _: u16) callconv(.c) i32 {
    return 1;
}
fn fakeSidAcquire() callconv(.c) i32 {
    return 11;
}
fn fakeSidWrite(_: u32, _: u8, _: u8) callconv(.c) i32 {
    return 0;
}
fn fakeSidRelease(_: u32) callconv(.c) i32 {
    return 0;
}
fn fakeMidiOpen(_: [*:0]const u8) callconv(.c) i32 {
    return 12;
}
fn fakeMidiSend(_: u32, _: u8, _: u8, _: u8, _: u8) callconv(.c) i32 {
    return 0;
}
fn fakeMidiClose(_: u32) callconv(.c) i32 {
    return 0;
}
fn fakeOplWrite(_: u8, _: u8, _: u8) callconv(.c) i32 {
    return 0;
}
fn fakeOplSimple() callconv(.c) i32 {
    return 0;
}

fn fakeInventorySummary(out: *r4os.abi.DeviceInventorySummary) callconv(.c) i32 {
    out.* = .{ .total = 2, .with_driver = 1, .without_driver = 1 };
    return 1;
}
fn fakeInventoryRecord(index: u32, out: *r4os.abi.DeviceInventoryRecord) callconv(.c) i32 {
    out.* = .{ .bus = @intCast(index + 1) };
    return 1;
}
fn fakeMemorySummary(out: *r4os.abi.ProgramMemorySummary) callconv(.c) i32 {
    out.* = .{ .physical_bytes = 64 * 1024 * 1024 };
    return 0;
}
fn fakeMemoryCount() callconv(.c) u32 {
    return 3;
}
fn fakeMemoryBlock(index: u32, out: *r4os.abi.ProgramMemoryBlockInfo) callconv(.c) i32 {
    out.* = .{ .id = index };
    return 1;
}
fn fakePressure(out: *r4os.abi.ProgramMemoryPressureSnapshot) callconv(.c) i32 {
    out.* = .{ .pressure_level = r4os.abi.memory_pressure_level_normal };
    return 1;
}
fn fakePerformance(out: *r4os.abi.ProgramPerformanceSummary) callconv(.c) i32 {
    out.* = .{ .audio_stream_writes = 9 };
    return 1;
}
fn fakeProgramInstanceStorageLegacy(out: *r4os.abi.ProgramInstanceStorageSummary) callconv(.c) i32 {
    @memset(@as([*]u8, @ptrCast(out))[0..256], 0);
    out.version = 1;
    out.size = 256;
    out.active_instances = 4;
    out.active_service_instances = 2;
    out.current_payload_bytes = 4096;
    out.payload_allocations = 11;
    out.payload_releases = 7;
    return 1;
}
fn fakeProgramInstanceStorageV2(out: *r4os.abi.ProgramInstanceStorageSummary) callconv(.c) i32 {
    if (out.version < 2 or out.size < @sizeOf(r4os.abi.ProgramInstanceStorageSummary)) return -1;
    out.* = .{ .active_instances = 4, .active_service_instances = 2, .current_payload_bytes = 4096, .payload_allocations = 11, .payload_releases = 7, .current_gui_frame_bytes = 8192 };
    return 1;
}
fn fakeProgramInstanceStorageSelfTest(out: *r4os.abi.ProgramInstanceStorageSelfTestResult) callconv(.c) i32 {
    out.* = .{ .cases = 27, .passed_cases = 27, .flags = 0xF, .allocation_failures_before = 2, .allocation_failures_after = 4 };
    return 1;
}
fn fakeExecutionInventory(out: *r4os.abi.ProgramInventorySummary) callconv(.c) i32 {
    out.* = .{
        .version = r4os.abi.program_inventory_version,
        .size = @sizeOf(r4os.abi.ProgramInventorySummary),
        .program_total = 17,
        .task_total = 29,
        .thread_total = 23,
        .program_peak = 41,
    };
    return r4os.abi.program_handle_ok;
}
fn fakePerformanceTask(index: u32, out: *r4os.abi.ProgramTaskPerformanceInfo) callconv(.c) i32 {
    out.* = .{ .id = index };
    return 1;
}
fn fakePerformanceStorage(index: u32, out: *r4os.abi.ProgramStoragePerformanceInfo) callconv(.c) i32 {
    out.* = .{ .index = index };
    return 1;
}
fn fakePerformanceBoot(_: u32, out: *r4os.abi.ProgramBootPhasePerformanceInfo) callconv(.c) i32 {
    out.* = .{};
    return 1;
}
fn fakeHardware(out: *r4os.abi.HardwareSummary) callconv(.c) i32 {
    out.* = .{ .cpu_logical_processors = 4 };
    return 1;
}

fn initFacades() struct { audio: r4os.Audio, devices: r4os.Devices } {
    sys_table = .{};
    sys_table.service_open = @intFromPtr(&fakeServiceOpen);
    sys_table.service_close = @intFromPtr(&fakeServiceClose);
    sys_table.service_call = @intFromPtr(&fakeServiceCall);
    audio_table = .{};
    audio_table.audio_open_stream = @intFromPtr(&fakeAudioOpen);
    audio_table.sid_acquire = @intFromPtr(&fakeSidAcquire);
    audio_table.sid_write_register = @intFromPtr(&fakeSidWrite);
    audio_table.sid_release = @intFromPtr(&fakeSidRelease);
    audio_table.midi_open_synth = @intFromPtr(&fakeMidiOpen);
    audio_table.midi_send = @intFromPtr(&fakeMidiSend);
    audio_table.midi_close = @intFromPtr(&fakeMidiClose);
    audio_table.opl3_write_register = @intFromPtr(&fakeOplWrite);
    audio_table.opl3_reset = @intFromPtr(&fakeOplSimple);
    audio_table.opl3_render_block = @intFromPtr(&fakeOplSimple);
    audio_table.opl3_stop = @intFromPtr(&fakeOplSimple);
    dev_table = .{};
    dev_table.device_inventory_summary = @intFromPtr(&fakeInventorySummary);
    dev_table.device_inventory_record = @intFromPtr(&fakeInventoryRecord);
    dev_table.memory_summary = @intFromPtr(&fakeMemorySummary);
    dev_table.memory_block_count = @intFromPtr(&fakeMemoryCount);
    dev_table.memory_block = @intFromPtr(&fakeMemoryBlock);
    dev_table.memory_pressure_snapshot = @intFromPtr(&fakePressure);
    dev_table.performance_summary = @intFromPtr(&fakePerformance);
    dev_table.program_instance_storage_summary = @intFromPtr(&fakeProgramInstanceStorageLegacy);
    dev_table.program_instance_storage_summary_v2 = @intFromPtr(&fakeProgramInstanceStorageV2);
    dev_table.program_instance_storage_self_test = @intFromPtr(&fakeProgramInstanceStorageSelfTest);
    dev_table.execution_inventory_summary = @intFromPtr(&fakeExecutionInventory);
    dev_table.performance_task = @intFromPtr(&fakePerformanceTask);
    dev_table.performance_storage = @intFromPtr(&fakePerformanceStorage);
    dev_table.performance_boot_phase = @intFromPtr(&fakePerformanceBoot);
    dev_table.hardware_summary = @intFromPtr(&fakeHardware);
    const bundle = r4os.program.Bundle{ .raw = &context, .sys = &sys_table, .audio = &audio_table, .dev = &dev_table };
    const sys = r4os.r4sys.Context.init(&bundle);
    return .{
        .audio = .{ .sys = sys, .raw = r4os.r4audio.Context.init(&bundle) },
        .devices = .{ .raw = r4os.r4dev.Context.init(&bundle) },
    };
}

test "PCM lifecycle validates format timeout progress and double close" {
    stream_open = false;
    open_connections = 0;
    timeout_next = false;
    var facades = initFacades();
    try std.testing.expect(facades.audio.available());
    const invalid_format: r4os.PcmFormat = @enumFromInt(2);
    try std.testing.expect(switch (facades.audio.openStream(48_000, 2, invalid_format, r4os.app_audio.default_volume, r4os.time_contract.timeoutPoll())) {
        .failure => |raw| raw == r4os.abi.service_api_result_invalid,
        else => false,
    });
    timeout_next = true;
    try std.testing.expect(switch (facades.audio.openStream(48_000, 2, .s16le, r4os.app_audio.default_volume, r4os.time_contract.timeoutPoll())) {
        .timed_out => true,
        else => false,
    });
    var stream = switch (facades.audio.openStream(48_000, 2, .s16le, r4os.app_audio.default_volume, r4os.time_contract.timeoutForever())) {
        .stream => |value| value,
        else => return error.Open,
    };
    var pcm: [2048]u8 = .{1} ** 2048;
    try std.testing.expectEqual(pcm.len, switch (stream.write(pcm[0..], r4os.time_contract.timeoutForever())) {
        .written => |bytes| bytes,
        else => 0,
    });
    try std.testing.expect(switch (stream.setVolume(0x8000, r4os.time_contract.timeoutForever())) {
        .ok => true,
        else => false,
    });
    try std.testing.expect(switch (stream.close(r4os.time_contract.timeoutForever())) {
        .ok => true,
        else => false,
    });
    try std.testing.expect(switch (stream.close(r4os.time_contract.timeoutForever())) {
        .failure => |raw| raw == r4os.abi.err_closed,
        else => false,
    });
    try std.testing.expectEqual(@as(u32, 0), open_connections);
}

test "advanced synth operations and R4DEV views stay explicit" {
    var facades = initFacades();
    const advanced = facades.audio.advanced();
    try std.testing.expectEqual(@as(i32, 11), advanced.sidAcquire());
    try std.testing.expectEqual(@as(i32, 0), advanced.sidWriteRegister(11, 0, 1));
    try std.testing.expectEqual(@as(i32, 12), advanced.midiOpenSynth("OPL3"));
    try std.testing.expectEqual(@as(i32, 0), advanced.midiSend(12, 0, 0x90, 60, 100));
    try std.testing.expectEqual(@as(i32, 0), advanced.opl3Reset());
    const inventory = facades.devices.inventory();
    try std.testing.expectEqual(@as(u32, 2), inventory.summary().?.total);
    try std.testing.expectEqual(@as(u32, 4), inventory.hardware().?.cpu_logical_processors);
    const memory = facades.devices.memory();
    try std.testing.expectEqual(@as(u64, 64 * 1024 * 1024), memory.summary().?.physical_bytes);
    try std.testing.expectEqual(@as(u32, 3), memory.blockCount());
    const performance = facades.devices.performance();
    try std.testing.expectEqual(@as(u64, 9), performance.summary().?.audio_stream_writes);
    const storage = performance.programInstanceStorage().?;
    try std.testing.expectEqual(@as(u32, 2), storage.version);
    try std.testing.expectEqual(@as(u32, @sizeOf(r4os.abi.ProgramInstanceStorageSummary)), storage.size);
    try std.testing.expectEqual(@as(u32, 4), storage.active_instances);
    try std.testing.expectEqual(@as(u64, 4096), storage.current_payload_bytes);
    try std.testing.expectEqual(@as(u64, 8192), storage.current_gui_frame_bytes);
    var legacy_storage: r4os.abi.ProgramInstanceStorageSummary = undefined;
    @memset(std.mem.asBytes(&legacy_storage), 0xA5);
    try std.testing.expect(performance.raw.programInstanceStorageSummaryLegacy(&legacy_storage) > 0);
    try std.testing.expectEqual(@as(u32, 1), legacy_storage.version);
    try std.testing.expectEqual(@as(u32, 256), legacy_storage.size);
    try std.testing.expectEqual(@as(u8, 0xA5), std.mem.asBytes(&legacy_storage)[256]);
    legacy_storage.version = 3;
    legacy_storage.size = @sizeOf(r4os.abi.ProgramInstanceStorageSummary);
    try std.testing.expect(performance.raw.programInstanceStorageSummaryV2(&legacy_storage) > 0);
    try std.testing.expectEqual(@as(u32, 2), legacy_storage.version);
    @memset(std.mem.asBytes(&legacy_storage), 0xB6);
    legacy_storage.version = 0;
    legacy_storage.size = @sizeOf(r4os.abi.ProgramInstanceStorageSummary) - 1;
    try std.testing.expectEqual(@as(i32, -1), performance.raw.programInstanceStorageSummaryV2(&legacy_storage));
    try std.testing.expectEqual(@as(u32, 0), legacy_storage.version);
    try std.testing.expectEqual(@as(u32, @sizeOf(r4os.abi.ProgramInstanceStorageSummary) - 1), legacy_storage.size);
    try std.testing.expectEqual(@as(u8, 0xB6), std.mem.asBytes(&legacy_storage)[8]);
    dev_table.size = 288;
    const legacy_fallback = performance.programInstanceStorage().?;
    try std.testing.expectEqual(@as(u32, 1), legacy_fallback.version);
    try std.testing.expectEqual(@as(u32, 256), legacy_fallback.size);
    try std.testing.expectEqual(@as(u64, 0), legacy_fallback.current_gui_frame_bytes);
    var storage_test: r4os.abi.ProgramInstanceStorageSelfTestResult = .{};
    try std.testing.expect(performance.programInstanceStorageSelfTest(&storage_test) > 0);
    try std.testing.expectEqual(@as(u32, 27), storage_test.passed_cases);
    try std.testing.expectEqual(@as(u32, 0xF), storage_test.flags);
    const execution = performance.executionInventory().?;
    try std.testing.expectEqual(r4os.abi.program_inventory_version, execution.version);
    try std.testing.expectEqual(@as(u32, 17), execution.program_total);
    try std.testing.expectEqual(@as(u32, 29), execution.task_total);
    try std.testing.expectEqual(@as(u32, 23), execution.thread_total);
    var direct_execution: r4os.abi.ProgramInventorySummary = .{};
    try std.testing.expectEqual(r4os.abi.program_handle_ok, facades.devices.raw.executionInventorySummary(&direct_execution));
    try std.testing.expectEqual(@as(u32, 41), direct_execution.program_peak);
}

File: Tests/Conformance/AudioDeviceFacadePrototype05827.c
#include <r4os/r4os.h>
#include <string.h>

static int timeout_next;
static int stream_open;
static uint32_t open_connections;

static int32_t fake_service_open(const uint8_t *name, R4ServiceInfo *out) {
    (void)name; *out = (R4ServiceInfo){0}; out->handle = 77u; ++open_connections; return 0;
}
static int32_t fake_service_close(uint32_t handle) {
    if (handle != 77u || open_connections == 0u) return R4OS_SERVICE_API_RESULT_BAD_HANDLE;
    --open_connections; return 0;
}
static int32_t fake_service_call(uint32_t handle, uint16_t op, const uint8_t *request, uint32_t request_len, R4ServiceMessageHeader *header, uint8_t *response, uint32_t capacity, uint64_t timeout) {
    (void)handle; (void)timeout;
    *header = (R4ServiceMessageHeader){0}; header->magic = R4OS_SERVICE_API_MAGIC; header->version = R4OS_SERVICE_API_VERSION; header->op = op; header->status = 0;
    if (timeout_next) { timeout_next = 0; return R4OS_SERVICE_API_RESULT_TIMEOUT; }
    if (op == R4OS_AUDIO_SERVICE_OP_MASTER_STATUS || op == R4OS_AUDIO_SERVICE_OP_SET_MASTER_STATE) {
        if (capacity < sizeof(R4AudioServiceMasterState)) return R4OS_SERVICE_API_RESULT_BUFFER_TOO_SMALL;
        R4AudioServiceMasterState state = {0}; state.magic = R4OS_AUDIO_MASTER_STATE_MAGIC; state.version = R4OS_AUDIO_MASTER_STATE_VERSION; state.size = sizeof(state); state.master_revision = 4u; state.selected_volume_fixed = 0x8000u; state.last_audible_volume_fixed = 0x8000u;
        if (op == R4OS_AUDIO_SERVICE_OP_SET_MASTER_STATE) {
            R4AudioServiceMasterRequest master;
            if (request_len != sizeof(master)) return R4OS_SERVICE_API_RESULT_INVALID;
            memcpy(&master, request, sizeof(master));
            if (master.magic != R4OS_AUDIO_MASTER_REQUEST_MAGIC || master.version != R4OS_AUDIO_MASTER_REQUEST_VERSION || master.size != sizeof(master)) return R4OS_SERVICE_API_RESULT_INVALID;
            if ((master.flags & R4OS_AUDIO_MASTER_REQUEST_FLAG_MUTED) != 0u) state.flags |= R4OS_AUDIO_MASTER_STATE_FLAG_MUTED;
        }
        memcpy(response, &state, sizeof(state)); return (int32_t)sizeof(state);
    }
    if (capacity < sizeof(R4AudioWireResult)) return R4OS_SERVICE_API_RESULT_BUFFER_TOO_SMALL;
    R4AudioWireResult result = {0}; result.magic = R4OS_AUDIO_SERVICE_RESULT_MAGIC; result.version = R4OS_AUDIO_SERVICE_RESULT_VERSION; result.action = op; result.stream_id = 91u;
    if (op == R4OS_AUDIO_SERVICE_OP_OPEN_STREAM) stream_open = 1;
    else if (op == R4OS_AUDIO_SERVICE_OP_WRITE_STREAM) {
        if (!stream_open || request_len < sizeof(R4AudioWireWriteRequest)) result.result = -1;
        else { R4AudioWireWriteRequest wire; memcpy(&wire, request, sizeof(wire)); result.bytes = wire.byte_count; result.result = (int32_t)wire.byte_count; }
    } else if (op == R4OS_AUDIO_SERVICE_OP_SET_STREAM_VOLUME) { if (!stream_open) result.result = -1; }
    else if (op == R4OS_AUDIO_SERVICE_OP_CLOSE_STREAM) { if (stream_open) stream_open = 0; else result.result = -1; }
    memcpy(response, &result, sizeof(result)); return (int32_t)sizeof(result);
}

static int32_t fake_audio_open(uint32_t rate, uint16_t channels, uint16_t format) { (void)rate; (void)channels; (void)format; return 1; }
static int32_t fake_sid_acquire(void) { return 11; }
static int32_t fake_sid_write(uint32_t handle, uint8_t reg, uint8_t value) { (void)handle; (void)reg; (void)value; return 0; }
static int32_t fake_sid_release(uint32_t handle) { (void)handle; return 0; }
static int32_t fake_midi_open(const uint8_t *backend) { (void)backend; return 12; }
static int32_t fake_midi_send(uint32_t handle, uint8_t channel, uint8_t status, uint8_t data1, uint8_t data2) { (void)handle; (void)channel; (void)status; (void)data1; (void)data2; return 0; }
static int32_t fake_midi_close(uint32_t handle) { (void)handle; return 0; }
static int32_t fake_opl_write(uint8_t bank, uint8_t reg, uint8_t value) { (void)bank; (void)reg; (void)value; return 0; }
static int32_t fake_opl_simple(void) { return 0; }

static int32_t fake_inventory_summary(R4DeviceInventorySummary *out) { *out = (R4DeviceInventorySummary){0}; out->total = 2u; out->with_driver = 1u; return 1; }
static int32_t fake_inventory_record(uint32_t index, R4DeviceInventoryRecord *out) { *out = (R4DeviceInventoryRecord){0}; out->bus = (uint8_t)(index + 1u); return 1; }
static int32_t fake_memory_summary(R4ProgramMemorySummary *out) { *out = (R4ProgramMemorySummary){0}; out->physical_bytes = 64u * 1024u * 1024u; return 0; }
static uint32_t fake_memory_count(void) { return 3u; }
static int32_t fake_memory_block(uint32_t index, R4ProgramMemoryBlockInfo *out) { *out = (R4ProgramMemoryBlockInfo){0}; out->id = index; return 1; }
static int32_t fake_pressure(R4ProgramMemoryPressureSnapshot *out) { *out = (R4ProgramMemoryPressureSnapshot){0}; out->pressure_level = R4OS_MEMORY_PRESSURE_LEVEL_NORMAL; return 1; }
static int32_t fake_performance(R4ProgramPerformanceSummary *out) { *out = (R4ProgramPerformanceSummary){0}; out->audio_stream_writes = 9u; return 1; }
static int32_t fake_program_instance_storage_legacy(R4ProgramInstanceStorageSummary *out) { memset(out, 0, 256u); out->version = 1u; out->size = 256u; out->active_instances = 4u; out->active_service_instances = 2u; out->current_payload_bytes = 4096u; out->payload_allocations = 11u; out->payload_releases = 7u; return 1; }
static int32_t fake_program_instance_storage_v2(R4ProgramInstanceStorageSummary *out) { if (out->version < 2u || out->size < sizeof(*out)) return -1; *out = (R4ProgramInstanceStorageSummary){0}; out->version = 2u; out->size = sizeof(*out); out->active_instances = 4u; out->active_service_instances = 2u; out->current_payload_bytes = 4096u; out->payload_allocations = 11u; out->payload_releases = 7u; out->current_gui_frame_bytes = 8192u; return 1; }
static int32_t fake_program_instance_storage_self_test(R4ProgramInstanceStorageSelfTestResult *out) { *out = (R4ProgramInstanceStorageSelfTestResult){0}; out->version = 1u; out->size = sizeof(*out); out->cases = 27u; out->passed_cases = 27u; out->flags = 0xFu; out->allocation_failures_before = 2u; out->allocation_failures_after = 4u; return 1; }
static int32_t fake_execution_inventory(R4ProgramInventorySummary *out) { *out = (R4ProgramInventorySummary){0}; out->version = R4OS_PROGRAM_INVENTORY_VERSION; out->size = sizeof(*out); out->program_total = 17u; out->task_total = 29u; out->thread_total = 23u; out->program_peak = 41u; return R4OS_PROGRAM_HANDLE_OK; }
static int32_t fake_task(uint32_t index, R4ProgramTaskPerformanceInfo *out) { *out = (R4ProgramTaskPerformanceInfo){0}; out->id = index; return 1; }
static int32_t fake_storage(uint32_t index, R4ProgramStoragePerformanceInfo *out) { *out = (R4ProgramStoragePerformanceInfo){0}; out->index = index; return 1; }
static int32_t fake_boot(uint32_t index, R4ProgramBootPhasePerformanceInfo *out) { (void)index; *out = (R4ProgramBootPhasePerformanceInfo){0}; return 1; }
static int32_t fake_hardware(R4HardwareSummary *out) { *out = (R4HardwareSummary){0}; out->cpu_logical_processors = 4u; return 1; }

static R4Timeout poll_timeout(void) { R4Timeout value = {0}; value.kind = R4OS_TIMEOUT_KIND_POLL; return value; }
static R4Timeout forever_timeout(void) { R4Timeout value = {0}; value.kind = R4OS_TIMEOUT_KIND_FOREVER; return value; }

static void init_app(R4App *app, R4XStartR4Sys *sys, R4XStartR4Audio *audio, R4XStartR4Dev *dev) {
    *sys = (R4XStartR4Sys){0}; sys->service_open = (uintptr_t)&fake_service_open; sys->service_close = (uintptr_t)&fake_service_close; sys->service_call = (uintptr_t)&fake_service_call;
    *audio = (R4XStartR4Audio){0}; audio->audio_open_stream = (uintptr_t)&fake_audio_open; audio->sid_acquire = (uintptr_t)&fake_sid_acquire; audio->sid_write_register = (uintptr_t)&fake_sid_write; audio->sid_release = (uintptr_t)&fake_sid_release; audio->midi_open_synth = (uintptr_t)&fake_midi_open; audio->midi_send = (uintptr_t)&fake_midi_send; audio->midi_close = (uintptr_t)&fake_midi_close; audio->opl3_write_register = (uintptr_t)&fake_opl_write; audio->opl3_reset = (uintptr_t)&fake_opl_simple; audio->opl3_render_block = (uintptr_t)&fake_opl_simple; audio->opl3_stop = (uintptr_t)&fake_opl_simple;
    *dev = (R4XStartR4Dev){0}; dev->size = R4XSTART_R4DEV_SIZE; dev->device_inventory_summary = (uintptr_t)&fake_inventory_summary; dev->device_inventory_record = (uintptr_t)&fake_inventory_record; dev->memory_summary = (uintptr_t)&fake_memory_summary; dev->memory_block_count = (uintptr_t)&fake_memory_count; dev->memory_block = (uintptr_t)&fake_memory_block; dev->memory_pressure_snapshot = (uintptr_t)&fake_pressure; dev->performance_summary = (uintptr_t)&fake_performance; dev->performance_task = (uintptr_t)&fake_task; dev->performance_storage = (uintptr_t)&fake_storage; dev->performance_boot_phase = (uintptr_t)&fake_boot; dev->hardware_summary = (uintptr_t)&fake_hardware; dev->program_instance_storage_summary = (uintptr_t)&fake_program_instance_storage_legacy; dev->program_instance_storage_summary_v2 = (uintptr_t)&fake_program_instance_storage_v2; dev->program_instance_storage_self_test = (uintptr_t)&fake_program_instance_storage_self_test; dev->execution_inventory_summary = (uintptr_t)&fake_execution_inventory;
    *app = (R4App){0}; app->system.table = sys; app->audio.table = audio; app->devices.table = dev;
}

int main(void) {
    R4App app; R4XStartR4Sys sys; R4XStartR4Audio raw_audio; R4XStartR4Dev raw_dev;
    init_app(&app, &sys, &raw_audio, &raw_dev); stream_open = 0; open_connections = 0u;
    R4AudioFacade audio = r4_app_audio(&app); R4AudioStream stream;
    if (!r4_audio_available(&audio)) return 1;
    R4AudioServiceMasterState master;
    if (r4_audio_master_status(&audio, forever_timeout(), &master).kind != R4_AUDIO_RESULT_OK || master.selected_volume_fixed != 0x8000u) return 21;
    if (r4_audio_set_master_state(&audio, R4OS_AUDIO_MASTER_REQUEST_FLAG_SET_VOLUME | R4OS_AUDIO_MASTER_REQUEST_FLAG_SET_MUTED, 0x8000u, 1, master.master_revision, forever_timeout(), &master).kind != R4_AUDIO_RESULT_OK || (master.flags & R4OS_AUDIO_MASTER_STATE_FLAG_MUTED) == 0u) return 22;
    if (r4_audio_open_stream(&audio, 48000u, 2u, 2u, 0x10000u, poll_timeout(), &stream).kind != R4_AUDIO_RESULT_FAILED) return 2;
    timeout_next = 1; if (r4_audio_open_stream(&audio, 48000u, 2u, R4_AUDIO_FORMAT_S16LE, 0x10000u, poll_timeout(), &stream).kind != R4_AUDIO_RESULT_TIMED_OUT) return 3;
    if (r4_audio_open_stream(&audio, 48000u, 2u, R4_AUDIO_FORMAT_S16LE, 0x10000u, forever_timeout(), &stream).kind != R4_AUDIO_RESULT_OK) return 4;
    uint8_t pcm[2048]; memset(pcm, 1, sizeof(pcm)); R4AudioResult wrote = r4_audio_stream_write(&stream, pcm, sizeof(pcm), forever_timeout());
    if (wrote.kind != R4_AUDIO_RESULT_OK || wrote.bytes != sizeof(pcm)) return 5;
    if (r4_audio_stream_set_volume(&stream, 0x8000u, forever_timeout()).kind != R4_AUDIO_RESULT_OK) return 6;
    if (r4_audio_stream_close(&stream, forever_timeout()).kind != R4_AUDIO_RESULT_OK || r4_audio_stream_close(&stream, forever_timeout()).raw_code != R4OS_ERR_CLOSED || open_connections != 0u) return 7;
    R4AdvancedAudio advanced = r4_audio_advanced(audio);
    if (r4_audio_sid_acquire(&advanced) != 11 || r4_audio_sid_write_register(&advanced, 11u, 0u, 1u) != 0 || r4_audio_midi_open_synth(&advanced, (const uint8_t *)"OPL3") != 12 || r4_audio_midi_send(&advanced, 12u, 0u, 0x90u, 60u, 100u) != 0 || r4_audio_opl3_reset(&advanced) != 0) return 8;

    R4Devices devices = r4_app_devices(&app); if (!r4_devices_available(&devices)) return 9;
    R4DeviceInventoryView inventory = r4_devices_inventory(devices); R4DeviceInventorySummary inv; R4HardwareSummary hardware;
    if (r4_device_inventory_summary(&inventory, &inv) <= 0 || inv.total != 2u || r4_device_hardware_summary(&inventory, &hardware) <= 0 || hardware.cpu_logical_processors != 4u) return 10;
    R4MemoryView memory = r4_devices_memory(devices); R4ProgramMemorySummary mem;
    if (r4_memory_summary(&memory, &mem) != 0 || mem.physical_bytes != 64u * 1024u * 1024u || r4_memory_block_count(&memory) != 3u) return 11;
    R4PerformanceView performance = r4_devices_performance(devices); R4ProgramPerformanceSummary perf;
    if (r4_performance_summary(&performance, &perf) <= 0 || perf.audio_stream_writes != 9u) return 12;
    R4ProgramInstanceStorageSummary instance_storage; R4ProgramInstanceStorageSelfTestResult storage_test;
    if (r4_program_instance_storage_summary(&performance, &instance_storage) <= 0 || instance_storage.version != 2u || instance_storage.size != sizeof(instance_storage) || instance_storage.active_instances != 4u || instance_storage.current_payload_bytes != 4096u || instance_storage.current_gui_frame_bytes != 8192u) return 13;
    if (r4_program_instance_storage_self_test(&performance, &storage_test) <= 0 || storage_test.cases != 27u || storage_test.passed_cases != 27u || storage_test.flags != 0xFu || storage_test.allocation_failures_after <= storage_test.allocation_failures_before) return 14;
    R4ProgramInventorySummary execution;
    if (r4_execution_inventory_summary(&performance, &execution) != R4OS_PROGRAM_HANDLE_OK || execution.version != R4OS_PROGRAM_INVENTORY_VERSION || execution.program_total != 17u || execution.task_total != 29u || execution.thread_total != 23u) return 15;
    if (r4dev_execution_inventory_summary(&app.devices, &execution) != R4OS_PROGRAM_HANDLE_OK || execution.program_peak != 41u) return 16;
    memset(&instance_storage, 0xA5, sizeof(instance_storage));
    if (r4dev_program_instance_storage_summary_legacy(&app.devices, &instance_storage) <= 0 || instance_storage.version != 1u || instance_storage.size != 256u || ((uint8_t *)&instance_storage)[256] != 0xA5u) return 17;
    instance_storage.version = 3u; instance_storage.size = sizeof(instance_storage);
    if (r4dev_program_instance_storage_summary_v2(&app.devices, &instance_storage) <= 0 || instance_storage.version != 2u || instance_storage.size != sizeof(instance_storage)) return 18;
    memset(&instance_storage, 0xB6, sizeof(instance_storage)); instance_storage.version = 0u; instance_storage.size = sizeof(instance_storage) - 1u;
    if (r4dev_program_instance_storage_summary_v2(&app.devices, &instance_storage) != -1 || instance_storage.version != 0u || instance_storage.size != sizeof(instance_storage) - 1u || ((uint8_t *)&instance_storage)[8] != 0xB6u) return 19;
    raw_dev.size = 288u;
    if (r4_program_instance_storage_summary(&performance, &instance_storage) <= 0 || instance_storage.version != 1u || instance_storage.size != 256u || instance_storage.current_gui_frame_bytes != 0u) return 20;
    return 0;
}
